pkg/services/ngalert/schedule/schedule.go

```diff
@@ -280,7 +280,7 @@ func (sch *schedule) schedulePeriodic(ctx context.Context) error {
 			// waiting for all rule evaluation routines to stop
 			waitErr := dispatcherGroup.Wait()
 			// close the state manager and flush the state
-			sch.stateManager.Close(ctx)
+			sch.stateManager.Close()
```

Comment:
and change this context to have a timeout (perhaps 15 seconds)

Reply:
Yep, I agree that this is the right place to set the timeout, rather than inside `Close`

Reply:
Initially, I went exactly that route. However, I was not sure that the timeout was the safe way of doing that because in the case of big states we can end up with partially committed data, which is worse than potentially long-lasting stop.

Reply:
Even if we use `context.Background()` people will either use SIGKILL or have the service manager send a SIGKILL after timeout (i.e. `SendSIGKILL` in `systemd`). Can we estimate how much state we need to save to the database and do something based on that heuristic?

Reply:
Correct, but in that case, that problem would be on the user side (they do not stop the process gracefully). The duration of the flushing depends on many factors: database connectivity, number of states, the server performance etc. We can set it to a big enough number like 5 minutes but I think this will not make it better than no timeout at all.

Reply:
What I meant is that if we don't have a timeout a lot of people will either use SIGKILL or configure their service manager to send SIGKILL because to them Grafana is unresponsive, and so the outcome is the same: partial committed data?

What I am thinking is if we can make some assumptions about the size of the state that needs to be persisted on shutdown and decide whether we can commit it within the timeout or not? If there is too much data to save to the database could we just shutdown without committing the state at all?

One other option is that we could write all the state to a file on disk (much faster than doing a large number of consecutive database upserts?) and then when Grafana starts up we read this file from disk and write it back to the database, similar to a WAL?

Reply:
> If there is too much data to save to the database could we just shutdown without committing the state at all?

I actually think the same. It looks like we do not need to do that because we save the state every time we process a rule. I added this to our weekly agenda 😄

Reply:
Opened PR https://github.com/grafana/grafana/pull/55504 to get rid of it altogether